rholang/examples/linking/tests/LinkedList_test.rho

```diff
@@ -0,0 +1,265 @@
+    LinkedListFold, LinkedListReverse, LinkedListForeach, LinkedListMap, LinkedListGet,
+    LinkedListIndexOf, LinkedListLength, LinkedListRange, LinkedListConstructor
+  in {
+    LinkedListApi("EmptyList", EmptyList) | LinkedListApi("Head", LinkedListHead) |
+    LinkedListApi("Tail", LinkedListTail) | LinkedListApi("Prepend", LinkedListPrepend) |
+    LinkedListApi("Fold", LinkedListFold) | LinkedListApi("Reverse", LinkedListReverse) | 
+    LinkedListApi("Foreach", LinkedListForeach) | LinkedListApi("Map", LinkedListMap) | 
+    LinkedListApi("Get", LinkedListGet) | LinkedListApi("IndexOf", LinkedListIndexOf) | 
+    LinkedListApi("Length", LinkedListLength) | LinkedListApi("Range", LinkedListRange) |
+    LinkedListApi("LinkedList", LinkedListConstructor) |
+    for(
+        emptyList <- EmptyList; head <- LinkedListHead; tail <- LinkedListTail;
+        prepend <- LinkedListPrepend; fold <- LinkedListFold; reverse <- LinkedListReverse;
+        foreach <- LinkedListForeach; map <- LinkedListMap; get <- LinkedListGet; 
+        indexOf <- LinkedListIndexOf; length <- LinkedListLength; range <- LinkedListRange;
+        LinkedList <- LinkedListConstructor
+    ) {
+      new
+        emptyList_test, head_test, tail_test, prepend_test, fold_test,
+        reverse_test, foreach_test, map_test, get_test, indexOf_test,
+        length_test, range_test, LinkedList_test, and
+      in {
+        contract and(b1, b2, return) = {
+          match [b1, b2] with [true, true] => {
+            return(true)
+          }
+          _ => { return(false) }
+        } |
+        contract emptyList_test(_) = {
+          //nothing much to test here, should just be the empty list constant
+          match emptyList == emptyList with true => { "EmptyList: Pass".display("\n") }
+          _ => { "EmptyList: Fail".display("\n") }
+        } |
+        contract head_test(_) = {
+          new NonEmptyList in {
+            LinkedList(1, NonEmptyList) | for(list <- NonEmptyList) {
+              new result1, result2, answer1, answer2 in {
+                head(list, result1) | head(emptyList, result2) |
+                answer1(1) | answer2(Nil) |
+                for(r1 <- result1; r2 <- result2; a1 <- answer1; a2 <- answer2) {
+                  match [r1 == a1, r2 == a2] with [true, true] => {
+                    "Head: Pass".display("\n")
+                  }
+                  _ => { "Head: Fail".display("\n") }
+                }
+              }
+            }
+          }
+        } |
+        contract tail_test(_) = {
+          //return the tail of the list
+          new NonEmptyList in {
+            LinkedList(1, NonEmptyList) | for(tl <- NonEmptyList) {
+              prepend(2, tl, NonEmptyList) | for(list <- NonEmptyList) {
+                new result1, result2, answer1, answer2 in {
+                  tail(list, result1) | tail(emptyList, result2) |
+                  answer1(tl) | answer2(Nil) |
```

Comment:
Is this equivalent to `answer1(2)` ?

Reply:
No, it would be equivalent to `answer1([1, emptyList])`

Reply:
I see; is there a reason not write it that way?

Reply:
Similar argument to what we were talking about on line 81. `tl` is an arbitrary constant, so could have been any list and I just happened to choose a simple one because we don't have a nice testing infrastructure.